pkg/api/datasources.go

```diff
@@ -212,6 +218,9 @@ func (hs *HTTPServer) DeleteDataSourceByName(c *models.ReqContext) response.Resp
 	cmd := &models.DeleteDataSourceCommand{Name: name, OrgID: c.OrgId}
 	err := hs.DataSourcesService.DeleteDataSource(c.Req.Context(), cmd)
 	if err != nil {
+		if _, ok := err.(models.ErrDatasourceSecretsPlugin); ok {
+			return response.Error(500, "Failed to delete datasource: "+err.Error(), err)
```

Comment:
First of all, thank you for the review :-)

We were originally just planning on displaying the generic error message in the client and logging something more detailed in the server log, but when we demoed to AMG they specifically asked for us to send detailed messages to the client. Their reason is that users on their platform won't have access to the server log, and showing them a detailed error makes the customer support process easier.

Reply:
> Their reason is that users on their platform won't have access to the server log, and showing them a detailed error makes the customer support process easier.

I can understand they want that, but we need to decide whether we think it's a good idea, in general it's not something we do today. Will it work with all future secret storage plugin implementations? Rather I would like if we had a solution where each error message returned to the client would include a request identifier and also used when logging the actual error server side. Then client could use the request identifier when contacting the support.

Including @sakjur here since you've been looking into new error handling solution. Thoughts?

Reply:
We should probably not send the actual Go error to the client. In addition to possibly sending sensitive internals in plaintext (we often do not have control over the content of Go error messages), it'll also mean that we'll implicitly make a commitment to ensuring that those errors are somewhat legible.

This argument keeps popping up in favor of exposing internals to end users. I believe that there is room for structural improvement here by letting Grafana keep a log of certain errors that is reasonable to provide access to for admins. That would allow us to effectively combine the best of both worlds, at a cost. It's far from ideal, but it makes sense to have some kind of admin log available to admins even if they don't have access to the server logs.

In this specific case, short-term, could we use the fact that we have an error type to provide an optional public message from the plugin so that it can say "this is fine to let the end user know"? That'll allow carefully (or not, the plugin could choose to be explicitly careless here) setting up errors that are _meant_ for the end user's eyes rather than reusing internal errors. That should strike a balance between optimizing for customer support for hosted Grafana solutions and keeping internals internal. It'll require more work, yes, but we can be intentional about what we share with who. I think that's worth something.

As for request identifier: We have the traceID attached to errors in the logs already. It's however quite big and without mnemonics, so maybe that's not the solution we're looking for here? I agree that that's something that'd be useful, access control errors used to have a unique identifier and I found it to be quite convenient.

Reply:
@sakjur Let me see if I am interpreting what you are suggesting correctly.

Right now with my impl, if there is _any_ error in the plugin, we simply return an error from our method and show it in the client.

It sounds like you are suggesting that we should add explicit handling for different exception types in the plugin. If the error is one of the types that we handle, instead of returning an error, we should actually return a payload, perhaps with a "humanReadableError" field filled in with something _we_ have populated.

If that is what you are suggesting, I like it! Would stop go errors from showing up in the client, but we would still show something more helpful than "failed to update".

Reply:
Pushed a commit that makes use of a change here https://github.com/grafana/grafana/pull/50722 that allows plugins to provide a user-friendly error messages when things go wrong. Technically there is nothing stopping devs from simply sending in "ugly" error messages as is, but the variable naming throughout would mean that is a deliberate choice that could come with risk.

Reply:
That’s absolutely fine, our job is more making good practices easy than making bad ones impossible 😄 There is the occasional reason to do something that might otherwise seem unwise. Excited to take a look tomorrow 🙂